src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.ServiceHub.Client;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
+    {
+        private class ConnectionPools
+        {
+            private readonly HubClient _hubClient;
+            private readonly HostGroup _hostGroup;
+            private readonly TimeSpan _timeout;
+
+            private readonly ReaderWriterLockSlim _shutdownLock;
+            private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
+
+            private readonly int _maxPoolConnections;
+
+            // keyed to serviceName. each connection is for specific service such as CodeAnalysisService
+            private readonly ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>> _pools;
+
+            // indicate whether pool should be used.
+            // it is mutable since it will set to false when this pool got shutdown
+            private volatile bool _usePool;
+
+            public ConnectionPools(
+                HubClient hubClient,
+                HostGroup hostGroup,
+                bool usePool,
+                int maxPoolConnection,
+                TimeSpan timeout,
+                ReferenceCountedDisposable<RemotableDataJsonRpc> remotableDataRpc)
+            {
+                _hubClient = hubClient;
+                _hostGroup = hostGroup;
+                _timeout = timeout;
+                _remotableDataRpc = remotableDataRpc;
+
+                _maxPoolConnections = maxPoolConnection;
+
+                // we have 4 services. so start from 4. later if we add more services, it will still work.
+                _pools = new ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>>(concurrencyLevel: 4, capacity: 4);
+
+                _usePool = usePool;
+                _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+            }
+
+            public Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
```

Comment:
💭 `callbackTarget` should be removed from this signature, as it does not apply to connection pooling scenarios